core/slice.go

```diff
@@ -354,33 +348,178 @@ func (sl *Slice) Append(header *types.WorkObject, domTerminus common.Hash, domOr
 				}
 			}
 
-			// Compute the cubic flow discount
-			discountedConversionAmount := misc.ApplyCubicDiscount(conversionAmountInQuai, header.ConversionFlowAmount())
-			discountedConversionAmountInInt, _ := discountedConversionAmount.Int(nil)
+			// sort the newInboundEtxs based on the decreasing order of the max slips
+			sort.SliceStable(newInboundEtxs, func(i, j int) bool {
+				etxi := newInboundEtxs[i]
+				etxj := newInboundEtxs[j]
+
+				var slipi *big.Int
+				if etxi.EtxType() == types.ConversionType {
+					// If no slip is mentioned it is set to 90%
+					slipAmount := new(big.Int).Set(params.MaxSlip)
+					if len(etxi.Data()) > 1 {
+						slipAmount = new(big.Int).SetBytes(etxi.Data()[:2])
+						if slipAmount.Cmp(params.MaxSlip) > 0 {
+							slipAmount = new(big.Int).Set(params.MaxSlip)
+						}
+						if slipAmount.Cmp(params.MinSlip) < 0 {
+							slipAmount = new(big.Int).Set(params.MinSlip)
+						}
+					}
+					slipi = slipAmount
+				} else {
+					slipi = big.NewInt(0)
+				}
+
+				var slipj *big.Int
+				if etxj.EtxType() == types.ConversionType {
+					// If no slip is mentioned it is set to 90%
+					slipAmount := new(big.Int).Set(params.MaxSlip)
+					if len(etxj.Data()) > 1 {
+						slipAmount = new(big.Int).SetBytes(etxj.Data()[:2])
+						if slipAmount.Cmp(params.MaxSlip) > 0 {
+							slipAmount = new(big.Int).Set(params.MaxSlip)
+						}
+						if slipAmount.Cmp(params.MinSlip) < 0 {
+							slipAmount = new(big.Int).Set(params.MinSlip)
+						}
+					}
+					slipj = slipAmount
+				} else {
+					slipj = big.NewInt(0)
+				}
+
+				return slipi.Cmp(slipj) > 0
+			})
 
-			// Apply K Quai discount before applying the quadratic discount, conversionAmount = (1-kQuaiDiscount)*conversionAmountInHash
 			kQuaiDiscount := header.KQuaiDiscount()
-			conversionAmountAfterKQuaiDiscount := new(big.Int).Mul(discountedConversionAmountInInt, new(big.Int).Sub(big.NewInt(params.KQuaiDiscountMultiplier), kQuaiDiscount))
-			conversionAmountAfterKQuaiDiscount = new(big.Int).Div(conversionAmountAfterKQuaiDiscount, big.NewInt(params.KQuaiDiscountMultiplier))
 
 			// stores the original etx values before the conversion so that it can be used once the
 			// new exchange rate is calculated
 			etxValuesBeforeConversion := make([]*big.Int, len(newInboundEtxs))
+			originalEtxValues := make([]*big.Int, len(newInboundEtxs))
 
+			actualConversionAmountInQuai := big.NewInt(0)
 			realizedConversionAmountInQuai := big.NewInt(0)
 
 			for i, etx := range newInboundEtxs {
-
-				originalValue := new(big.Int).Set(etx.Value())
-
 				// If the etx is conversion
 				if types.IsConversionTx(etx) {
 					value := new(big.Int).Set(etx.Value())
```

Comment:
Can you check the .Sign() here to make sure it's 1 (a positive number)? Very simple sanity check but could save us if something goes wrong